Apply pattern to the post when selected

renatho opened the pull request:
Fixes #5140
Fixes #5130
Fixes #5133

### Changes proposed in this Pull Request

* This PR has some general fixes (sorry for that in this PR), but the main goal is to apply the pattern correctly when it's chosen.
* It also finalizes the pattern steps for course and lesson with the proper styles and content.

### Testing instructions

<!--
If the functionality of this PR is behind a feature flag, please include the
following testing step, using the correct name for the feature flag. (If you
aren't sure, just ignore this step)

* Enable the feature flag: `add_filter( 'sensei_feature_flag_{THE_FLAG_NAME}', '__return_true' );`
-->
* Create a new course.
* In the wizard form, add a title and a description.
* Navigate through the steps until the "Course Layout".
* Click in a pattern.
* Make sure it's applied to the course with the title and description replaced, and the course is saved.
* Repeat the test, but close the modal instead of choosing a pattern.
* Make sure the modal is closed, the default template is used in the course, and the course is saved.
* Also make sure that the pattern selection work on a new lesson.

<!--
Helpful tips for screenshots:
https://en.support.wordpress.com/make-a-screenshot/
-->
### Screenshot / Video

https://user-images.githubusercontent.com/876340/170561649-2ce9480c-d3b8-4de7-aa90-2d9696f6a4ac.mov

renatho committed:
Fix footer spacing

renatho committed:
Update logic to set pattern after chosen

renatho committed:
Save post on wizard completion

renatho committed:
Add class to sensei title and description

renatho committed:
Fill the patterns content with user input

renatho committed:
Unify patterns step

renatho committed:
Update close hover color because the background on the patterns is white

renatho committed:
Fix wizard footer

renatho committed:
Set title sticky on the patterns step

renatho committed:
Remove paragraph from Patterns step

renatho committed:
Add steps counter translation

renatho committed:
Simplify return

renatho committed:
Add custom title to lesson and course steps

renatho committed:
Refactor selection responsability

renatho committed:
Simplify return

renatho committed:
Refactor placeholders replacement

renatho commented on assets/admin/editor-wizard/steps/course-patterns-step.js:
```diff
@@ -1,43 +1,39 @@
 
-	// We can call `onCompletion` to complete the wizard after setting the correct pattern with `setData`.
-	// We could replace `onCompletion` with the `goToNextStep` callback with a similar effect.
+	if ( data.courseTitle ) {
```
When we implement the final patterns, if we don't have a course title to replace, we can remove this condition.

renatho committed:
Add tests for pattern placeholders replace

renatho committed:
Add tests to replace placeholders

aaronfc commented on assets/admin/editor-wizard/steps/lesson-patterns-step.js:
```diff
@@ -1,43 +1,22 @@
-	);
-};
+const LessonPatternsStep = ( props ) => (
+	<PatternsStep title={ __( 'Lesson Type', 'sensei-lms' ) } { ...props } />
```
I guess we will add replacements here for the title (if finally are needed).

renatho replied:
Sorry! Not sure if I got your point here. Could you explain more?

renatho replied:
Ah, reading again, I think it's the same you meant with this following sentence, right?

> Lesson step will need to be implemented after https://github.com/Automattic/sensei/pull/5194 is integrated, right?

aaronfc replied:
That's it! :D

aaronfc commented on assets/admin/editor-wizard/steps/patterns-step.test.js:
```diff
@@ -0,0 +1,72 @@
+ */
+import { replacePlaceholders } from './patterns-step';
+
+describe( 'replacePlaceholders', () => {
```
❤️

aaronfc commented on assets/admin/editor-wizard/wizard.js:
```diff
@@ -34,20 +36,25 @@ const Wizard = ( { steps, onCompletion } ) => {
 				<div className="sensei-editor-wizard__footer">
 					<div className="sensei-editor-wizard__progress">
-						Step { currentStepNumber + 1 } of { steps.length }
+						{ sprintf(
+							// translators: %1$d Current step number, %2$d Number of steps.
+							__( 'Step %1$d of %2$d', 'sensei-lms' ),
+							currentStepNumber + 1,
+							steps.length
+						) }
```
Nice catch! 😅

aaronfc commented on assets/admin/editor-wizard/wizard.js:
```diff
@@ -34,20 +36,25 @@ const Wizard = ( { steps, onCompletion } ) => {
 			</div>
-		) ) ||
-		null
+		)
```
I remember needing to add this `|| null` because I was getting an error that a render should return something always. Why it is not needed anymore?

renatho replied:
Ah, I got it! If `steps[ currentStepNumber ]` is `undefined` it would fire an error. But if we return `null` we also have a weird behavior with the modal open, but only with the overlay.

Thinking a little more, I don't think we need this check since `goToNextStep` will only increment if we have steps. I removed it for now in this commit: https://github.com/Automattic/sensei/pull/5179/commits/d64d74ee91981b2c376a32a9e95f23fcf1b16271, but let me know if I'm missing anything. 😉

renatho replied:
I also removed the conditional in the other case where steps should always be an array.

aaronfc reviewed (dismissed):
Looks good!
Just some clarifications:
- Lesson step will need to be implemented after [this](https://github.com/Automattic/sensei/pull/5194) is integrated, right?
- I saw that Course Outline is added with some default lessons, which is nice, but it is not when using the default template (course outline appears without lessons) – this will be fixed [here](https://github.com/Automattic/sensei/pull/5196), right?

renatho committed:
Remove conditional from component returns

renatho committed:
Merge branch 'feature/course-lesson-wizard' of github.com:Automattic/sensei into add/pattern-selection

renatho committed:
Add lesson title replacement

renatho commented:
Thank you for your review, @aaronfc !

> Lesson step will need to be implemented after https://github.com/Automattic/sensei/pull/5194 is integrated, right?

Definitely! Good catch! I already updated the branch, since that was merged, and added the implementation here: https://github.com/Automattic/sensei/pull/5179/commits/24207cad1dc9aebf884dc8cb8e6d509b054478d9

> I saw that Course Outline is added with some default lessons, which is nice, but it is not when using the default template (course outline appears without lessons) – this will be fixed https://github.com/Automattic/sensei/pull/5196, right?

That's correct! Also about the format of the Outline in the patterns, it's something temporary, I think it will depend on how we define them.

aaronfc reviewed (approved):
Looks good works well :D

alexsanford commented:
When I test this (on the feature branch), and either close the modal or click "Start with default layout" (rather than selecting a pattern), it seems like the description that I've added to the modal is lost (it's not added to the post at all). Is this intentional? Should we have some way of adding the description to the default template? Or are we ok with just losing the description if the modal is closed?

My thought is we should have the description in the default template somewhere.

cc @renatho @burtrw

renatho commented:
> it seems like the description that I've added to the modal is lost (it's not added to the post at all). Is this intentional?

Good point! For now, it was intentional, just applying the description to the patterns. And when skipping, it just loads the default template. But I think it makes sense. Maybe we could add a paragraph with the filled description.

If we go for that, I have a question if the user doesn't add any information to the description input. In the case of the patterns, we're keeping the original content of the pattern. Maybe should we also have a paragraph with an example text in the default template?

burtrw commented:
I was thinking the description could/should also get applied as the Course 'excerpt' if that is possible?

But adding a simple text block at the top of the default that includes the description would also be great.

We could either just put a blank paragraph block where the description would go or have default text. My concern with default text is translations (though we are adding default text to other patterns...). 

What do y'all think would be best?

renatho commented:
> I was thinking the description could/should also get applied as the Course 'excerpt' if that is possible?

Yep. It's possible to do it in a similar way to how we update the title currently.

> But adding a simple text block at the top of the default that includes the description would also be great.
> 
> We could either just put a blank paragraph block where the description would go or have default text. My concern with default text is translations (though we are adding default text to other patterns...).
> 
> What do y'all think would be best?

IMO, I think since it's the default template and not a styled pattern, we could just leave an empty paragraph if the user doesn't inform a description. We could have a placeholder there how we have in the current lesson editor _"Write lesson content..."_

<img width="739" alt="Screen Shot 2022-05-31 at 11 23 31" src="https://user-images.githubusercontent.com/876340/171197059-59c52951-7b7b-422c-8fcb-e41c16970ff7.png">

I wait for confirmation or a new suggestion to open an issue for that. 😉
